### PR TITLE
Fix `dhall freeze` to better handle protected imports

### PR DESCRIPTION
**Pull request by Gabriel439**
This includes two changes:

* `dhall freeze` will now fix an ordinary protected import that does not
  yet have a fallback import
* `dhall freeze` will now add a fallback import to an import that is
  already protected

### CODE REVIEW

**Comment by sjakobi** [comment]
@Gabriel439 Do the two changes from the PR description apply only with the `--cache` flag or in general?

**Comment by Gabriel439** [comment]
@sjakobi: These changes only affect the `--cache` flag.  The only thing that is modified is the `cache` function, which is only used to rewrite an expression if the `--cache` flag is enabled

File: dhall/src/Dhall/Freeze.hs [hunk 1]
```diff
@@ -158,15 +158,36 @@ freeze outputMode input scope intent characterSet censor = do
     let freezeFunction = freezeScope directory
 
     let cache
+            -- This case is necessary because `transformOf` is a bottom-up
+            -- rewrite rule.   Without this rule, if you were to transform a
```
Comment (sjakobi) [inline dhall/src/Dhall/Freeze.hs:162]: Tangential: Are there any top-down rewrite rules that we could use instead? I believe we have to apply similar workarounds in other places.

File: dhall/src/Dhall/Freeze.hs [hunk 2]
```diff
@@ -183,11 +204,26 @@ freeze outputMode input scope intent characterSet censor = do
...
                 -}
                 if frozenImport /= import_
                     then return (ImportAlt (Embed frozenImport) (Embed import_))
                     else return (Embed import_)
+        cache
+            (Embed import_@(Import { importHashed = ImportHashed { hash = Just _ } })) = do
+                -- Regenerate the integrity check, just in case it's wrong
+                frozenImport <- freezeFunction import_
+
+                -- `dhall freeze` also works the other way around, adding an
```
Comment (sjakobi) [inline dhall/src/Dhall/Freeze.hs:217]:
```suggestion
                -- `dhall freeze --cache` also works the other way around, adding an
```

?

File: dhall/src/Dhall/Freeze.hs (hunk 1 above)
Comment (sjakobi) [reply dhall/src/Dhall/Freeze.hs:162]: > Are there any top-down rewrite rules that we could use instead?

I assume the types probably wouldn't work out…

Comment (Gabriel439) [reply dhall/src/Dhall/Freeze.hs:162]: There aren't any that I know of.

**Comment by Gabriel439** [comment]
@sjakobi: I went ahead and added tests to confirm that this behaves as expected, so I'll go ahead and mege

**Comment by sjakobi** [comment]
Thank you for adding tests, @Gabriel439! :)

**Comment by Gabriel439** [comment]
@sjakobi: You're welcome! 🙂

**Pull request merged**